### PR TITLE
[DEV-287] - Change function visibility according to slither suggestions

### PR DESCRIPTION
**Pull request by grp06**
- Updated Solidity to version 0.8.9 according to Slither's suggestion:
`Pragma version 0.8.9 (src/Raft.sol#2) necessitates a version too recent to be trusted. Consider deploying with 0.6.12/0.7.6/0.8.7`
- change initializer visibility to `public` to give flexibility for future child contracts: https://forum.openzeppelin.com/t/slither-upgradeable-contract-does-not-protect-its-initialize-function/28432
- Organized and cleaned up unit tests.

## Unit test checklist
Cases to test for
- [x] give
	- [x] happy path
	- [x] give with bad token uri
	- [x] give with unauthorized sender
	- [x] give, give again
	- [x] give to self
	- [x] give with tokenId that already exists
- [x] take
	- [x] happy path
	- [x] take with bad token uri
	- [x] take with unauthorized sender
	- [x] take, take again
	- [x] take to self
	- [x] take with tokenId that already exists
 - [x] data holder
	 - [x] set data holder
	 - [x] set data holder as non owner
 - [x] create spec as raft owner
	 - [x] happy path
	 - [x] raft owner isn't message sender
	 - [x] already registered spec
 - [x] unequip
	 - [x] happy  path
	 - [x] non authorized unequip
	 - [x] non-existent tokenId
- [x] ownership
	- [x] transfer ownership
	- [x] transfer ownership as non-owner

### CODE REVIEW

File: src/Badges.sol [hunk 1]
```diff
@@ -116,26 +116,26 @@ contract Badges is
...
   }
 
-  function tokenURI(uint256 tokenId) public view virtual override returns (string memory) {
+  function tokenURI(uint256 tokenId) external view virtual override returns (string memory) {
     require(_exists(tokenId), "tokenURI: token doesn't exist");
     return _tokenURIs[tokenId];
   }
 
-  function unequip(uint256 tokenId) public virtual override {
+  function unequip(uint256 tokenId) external virtual override {
```
Comment (rahulrumalla) [inline src/Badges.sol:132]: Hmm.. I think some of these functions if marked as external, will not be available for other contracts to call them, no?

File: src/Badges.t.sol [hunk 2]
```diff
@@ -17,14 +17,6 @@ contract ERC1271Mock {
   constructor(bool pass_) {
     pass = pass_;
   }
-
-  function isValidSignature(bytes32 hash, bytes memory signature) public view returns (bytes4) {
```
Comment (rahulrumalla) [inline src/Badges.t.sol:21]: why are we removing this? I think this is needed to check for signatures originating from contracts (liek multisig).. did we remove this test coverage?

File: src/Raft.sol [hunk 3]
```diff
@@ -29,7 +29,7 @@ contract Raft is ERC721EnumerableUpgradeable, UUPSUpgradeable, OwnableUpgradeabl
     address nextOwner,
     string memory name_,
     string memory symbol_
-  ) external initializer {
+  ) public initializer {
```
Comment (rahulrumalla) [inline src/Raft.sol:32]: what's the general rationale here?

**Review by rahulrumalla — COMMENTED** [review]
Left some comments. Could you perhaps add all your notes for context in the PR, please? I'm lacking context on most of these to provide feedback.

File: src/Badges.sol [hunk 4]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: GPL-3.0-only
-pragma solidity 0.8.9;
+pragma solidity 0.8.7;
```
[the remark this answers is not in the document]
Comment (grp06) [reply src/Badges.sol:2]: Slither says `Pragma version0.8.9 (src/Raft.sol#2) necessitates a version too recent to be trusted. Consider deploying with 0.6.12/0.7.6/0.8.7`

File: src/Raft.sol (hunk 3 above)
Comment (grp06) [reply src/Raft.sol:32]: > You may also consider changing the initializer to `public` if you want the flexibility of calling that initializer from a child contract, for example.

https://forum.openzeppelin.com/t/slither-upgradeable-contract-does-not-protect-its-initialize-function/28432

File: src/Badges.t.sol (hunk 2 above)
Comment (grp06) [reply src/Badges.t.sol:21]: It was an unused function. It's also unused in the ERC4973 repo

File: src/Badges.sol (hunk 1 above)
Comment (grp06) [reply src/Badges.sol:132]: Other contracts could call it but if another contract inherits from this, that contract would not be able to call it internally. You're right, I'll change these back. We should leave that flexibility.

File: src/Badges.sol [hunk 5]
```diff
@@ -67,7 +67,7 @@ contract Badges is
     address to,
     string calldata uri,
     bytes calldata signature
-  ) external virtual returns (uint256) {
+  ) external virtual override returns (uint256) {
```
Comment (grp06) [inline src/Badges.sol:70]: Here it wasn't compiling without `override`, This function is defined in the interface, so we need `override`. I'm actually not sure how it was compiling before

**Comment by grp06** [comment]
- Updated Solidity to version 0.8.9 to be aligned with the latest version of Solidity that Hardhat is compatible with.
- change initializer visibility to `public` to give flexibility for future child contracts: https://forum.openzeppelin.com/t/slither-upgradeable-contract-does-not-protect-its-initialize-function/28432
- Organized and cleaned up unit tests.

## Unit test checklist
Cases to test for
- [ ] give
	- [x] happy path
	- [x] give with bad token uri
	- [x] give with unauthorized sender
	- [x] give, give again
	- [x] give to self
	- [x] give with tokenId that already exists
- [ ] take
	- [x] happy path
	- [x] take with bad token uri
	- [x] take with unauthorized sender
	- [x] take, take again
	- [x] take to self
	- [x] take with tokenId that already exists
 - [x] data holder
	 - [x] set data holder
	 - [x] set data holder as non owner
 - [x] create spec as raft owner
	 - [x] happy path
	 - [x] raft owner isn't message sender
	 - [x] already registered spec
 - [x] unequip
	 - [x] happy  path
	 - [x] non authorized unequip
	 - [x] non-existent tokenId
- [x] ownership
	- [x] transfer ownership
	- [x] transfer ownership as non-owner

**Review by rahulrumalla — APPROVED** [review]
LGTM!

**Pull request merged**